pkg/cgroups/cgroups.go

```diff
@@ -127,20 +127,19 @@ func CheckOrMountCgrpFS(mapRoot string, runsOnKind bool) {
 		if mapRoot == "" {
 			mapRoot = cgroupRoot
 		}
-		err := cgrpCheckOrMountLocation(mapRoot)
-		// Failed cgroup2 mount is not a fatal error, sockmap will be disabled however
-		if err == nil {
-			log.Infof("Mounted cgroupv2 filesystem at %s", mapRoot)
-
-			if runsOnKind {
-				p, err := getCgroupRootForKind()
-				if err != nil {
-					log.WithError(err).
-						Warnf("Failed to determine cgroup v2 hierarchy for Kind node. Socket-based LB (--%s) will not work.",
-							option.EnableHostReachableServices)
-				} else {
-					setCgroupRoot(p)
-				}
+		if err := cgrpCheckOrMountLocation(mapRoot); err != nil {
```

Comment:
In general, `Fatal()` should be avoided.

The current set of accepted uses for `Fatal()` in Cilium is fairly limited, mostly for when the user has explicitly configured a set of commandline options that is invalid.

As far as I can tell, we never required cgroups v2 before, and most environments do not have cgroups v2 enabled, and this function is unconditionally run. This seems dangerous, if someone upgrades Cilium to this new version it will cause Cilium to crash and force the user to somehow resolve the issue. Am I missing something?

Reply:
@joestringer, would it be better just notify the user that socket-lb will disabled and the reason (failure to mount cgroupv2)?

Reply:
I think it would be like this, right?:
https://github.com/cilium/cilium/blob/819942ea1cbdc4b88741ba0b27a36a83941129b8/pkg/cgroups/cgroups.go#L136-L140

Reply:
We do need cgroupv2 fs mounted in order for socket-lb to work. This has always been the case. So if kube-proxy replacement is enabled, I would suggest that we fail early in cilium agent.

Reply:
> As far as I can tell, we never required cgroups v2 before, and most environments do not have cgroups v2 enabled, and this function is unconditionally run. This seems dangerous, if someone upgrades Cilium to this new version it will cause Cilium to crash and force the user to somehow resolve the issue. Am I missing something?

I clearly understand what you mean, @joestringer, and I understand the concern, but I wonder what @brb thinks based on his [note](https://github.com/cilium/cilium/pull/15999#issuecomment-848609493):

> I think we should not ignore the error. The comment in the code says:
>        // Failed cgroup2 mount is not a fatal error, sockmap will be disabled however
> We cannot no longer assume this, as bpf_sock (aka socket-lb) also depends on the mount.

I know how simple changes like these can demand a deeper analysis so, please, let us know how to help in this matter.

Edit: @aditighag have already exposed more information. Tks.

Reply:
Thanks for the context @aditighag , I wasn't thinking about the particular options that may require it.

I still think that Fatal is a bad idea. If we can perform a pure check of e.g. commandline args then I think it's reasonable to fatal out so that we can inform the user quickly and let them change the inputs they provide to Cilium. This will typically only occur upon first setup / install of Cilium. If a check involves the state of the system itself, then Cilium upgrades or system restart (could involve kernel upgrade) could then randomly trigger Cilium to stop running via the `Fatal()`. Cilium then is harder to debug since it's not running and we can't gather state from it during runtime, and it also increases the impact of issues like #14333. I'd say there should be a pretty small exclusive set of system state checks that could lead to `Fatal()`, primarily the baseline BPF availability check.

If there are options that require cgroup mounting and it cannot be mounted, I would suggest logging at Error level since Cilium is being asked to implement some functionality (eg socket LB) and it cannot implement that functionality, and furthermore it would not work around the issue. If there is a runtime workaround, then maybe issuing a Warning is sufficient.

Reply:
@joestringer I have mixed feelings about this. If users have enabled socket-based functionality, and cilium is unable to provide that, how can we convey that more prominently? More importantly, if Cilium wasn't able to mount cgroup 2 fs, won't `init.sh` fail to attach the BPF cgroup programs anyway? That'll lead to agent pod crashing, no? This could be worse in terms of debugging.

Reply:
It's a good question, but we really need to be careful that we don't make problems worse for users by crashing out the Cilium-agent pod when it could otherwise continue to operate without trouble. If we're not careful, we can cause more problems with the message than what the underling failure causes.

If socket LB is not effective but we're falling back to per-packet LB, then we are not breaking datapath functionality in general, we're gracefully degrading. We must not fatal out if we are gracefully degrading functionality. If the user isn't paying attention to the log messages and they're not noticing a problem, then is there even really a problem?

> More importantly, if Cilium wasn't able to mount cgroup 2 fs, won't init.sh fail to attach the BPF cgroup programs anyway? That'll lead to agent pod crashing, no? This could be worse in terms of debugging.

It looks like this is the case iff socket LB is enabled. Currently we mount cgroupv2 unconditionally. So one example that doesn't fit well here is if you deploy with socket LB / KPR disabled and happen to hit a cgroupv2 mount issue. In this situation, a warning message would allow Cilium to continue uninhibited. With a Fatal, we would break pod deployment, maybe even break L7 traffic forwarding (ref. #14333). Sure the user may notice more quickly but this can cause an outage rather than business as usual + investigate a weird log message later on.

In terms of debugging, while the final error message that states why the fatal occurred in this scenario would not include the failure to mount cgroupfs, it should still be clear just a couple of lines up in the logs so I think that's sufficient to still debug and understand the root cause pretty quickly.

Reply:
>It looks like this is the case iff socket LB is enabled.

That's why I [suggested](https://github.com/cilium/cilium/pull/15999#discussion_r646973082) to only do a fatal log when kpr/host-reachable-svc is enabled. :) Regardless of fatal or not, we should log a warning only when these configs are enabled.

Reply:
Hi Folks, I update the code inserting a check for `EnableHostReachableServices == true`.
Now the warning will be printed only for this condition ;)

Reply:
Keep in mind that the cgroupv2 is required for the sockops too. Also, if we fatal, then we need to check whether kpr != probe. As in the latter case, we can just disable the socket-lb.